### PR TITLE
[MLP-157] Added abstracted data provider, services and models examples

### PR DESCRIPTION
**Pull request by jsmuguruza**
## Background

As we discussed with @mvazquez-ms the idea was to create an example of how to abstract the data providers with their data models and use them in services in a running example. 

It's quite simple, but there are some things on generics that we couldn't make work as we really wanted to with Marce.

## Changes done

1. Modified 2nd storyboard to have a TableView and CustomCell.
2. Created new TableViewController and TableViewCell.
3. Created a Service to consume Data Provider.
4. Created a Data Provider Protocol to abstract it's functionality then implemented it with a simple example.

## Approvers required

@mvazquez-ms @lpelizzaMS @bsztamfater-ms  @mmaldini33 

## Additional Approvers

@AlphaGit  @andresmoschini  @matiasbeckerle  @mravinale

### CODE REVIEW

File: IOS/SwiftSeedProject/SwiftSeedProject/Domain/Services/PersonService.swift [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+//  Created by Mugu on 3/28/17.
+//  Copyright © 2017 Making Sense. All rights reserved.
+//
+
+import Foundation
+
+open class PersonService {
+    let dataProvider : InMemoryDataProvider
+    
+    var persons = [Person]()
```
Comment (mmaldini33) [inline IOS/SwiftSeedProject/SwiftSeedProject/Domain/Services/PersonService.swift:14]: Type Annotation for Empty Arrays and Dictionaries

For empty arrays and dictionaries, use type annotation. (For an array or dictionary assigned to a large, multi-line literal, use type annotation.)

Preferred:
var names: [Person] = []

Not Preferred:
var persons = [Person] ()

**Review by mmaldini33 — COMMENTED** [review]
Minimal comment.

File: IOS/SwiftSeedProject/SwiftSeedProject/Domain/Models/Person.swift [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+//  SwiftSeedProject
+//
+//  Created by Mugu on 3/28/17.
+//  Copyright © 2017 Making Sense. All rights reserved.
+//
+
+import Foundation
+
+class Person : NSObject {
+    var name: String!
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Domain/Models/Person.swift:12]: Please, add a blank line between these two.

File: IOS/SwiftSeedProject/SwiftSeedProject/Domain/Services/PersonService.swift [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+//  SwiftSeedProject
+//
+//  Created by Mugu on 3/28/17.
+//  Copyright © 2017 Making Sense. All rights reserved.
+//
+
+import Foundation
+
+open class PersonService {
+    let dataProvider : InMemoryDataProvider
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Domain/Services/PersonService.swift:12]: Please, add a blank line between these two.

Comment (mvazquez-ms) [reply IOS/SwiftSeedProject/SwiftSeedProject/Domain/Services/PersonService.swift:12]: `dataProvider` should be declared as

`let dataProvider : DataProvider`

(As the protocol, not the implementation)

File: IOS/SwiftSeedProject/SwiftSeedProject/Domain/Services/PersonService.swift [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+//  PersonService.swift
+//  SwiftSeedProject
+//
+//  Created by Mugu on 3/28/17.
+//  Copyright © 2017 Making Sense. All rights reserved.
+//
+
+import Foundation
+
+open class PersonService {
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Domain/Services/PersonService.swift:11]: Why is this class `open`?

File: IOS/SwiftSeedProject/SwiftSeedProject/Domain/Services/PersonService.swift [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    let dataProvider : InMemoryDataProvider
+    
+    var persons = [Person]()
+    
+    init() {
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Domain/Services/PersonService.swift:16]: Usually, the constructor should set dependencies and initial configuration, but nothing else.

The query to the DataProvider should be done separately, on a specific method or changing persons property to be lazy.

File: IOS/SwiftSeedProject/SwiftSeedProject/Domain/Services/PersonService.swift (hunk 1 above)
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Domain/Services/PersonService.swift:14]: This array should be declared as `let`.

Even if you modify it, you'll (and should) modify only the content, not the array itself (e.g. recreating it)

File: IOS/SwiftSeedProject/SwiftSeedProject/Domain/Services/PersonService.swift [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        dataProvider = InMemoryDataProvider()
+        let result : [AnyObject] = dataProvider.getEntities()
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Domain/Services/PersonService.swift:18]: `result` name is waaaaaaaaaaaaay too generic.

Name should be something like `personEntities` or similar.

Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Domain/Services/PersonService.swift:18]: `let result : [AnyObject] = dataProvider.getEntities()`

This is not taking advantage of generics usage, it should be written as

`let result : [PersonDataModel] = dataProvider.getEntities()`

File: IOS/SwiftSeedProject/SwiftSeedProject/Domain/Services/PersonService.swift [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+        
+        for object in result {
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Domain/Services/PersonService.swift:20]: Both, `object` and `result` are painfully generic names.

File: IOS/SwiftSeedProject/SwiftSeedProject/Domain/Services/PersonService.swift [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+            let person = Person(dataModel: object as! PersonDataModel)
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Domain/Services/PersonService.swift:21]: Casting here could (and should) be avoided if the change suggested on [this ](https://github.com/makingsensetraining/mobile-pocs/pull/13/files#r109415104)comment is implemented.

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Persistence/Implementations/InMemoryDataProvider.swift [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+//  SwiftSeedProject
+//
+//  Created by Mugu on 3/29/17.
+//  Copyright © 2017 Making Sense. All rights reserved.
+//
+
+import Foundation
+
+class InMemoryDataProvider  {
+    let entities : [AnyObject] = [
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Persistence/Implementations/InMemoryDataProvider.swift:12]: Please, add a blank line between these two.

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Persistence/Implementations/InMemoryDataProvider.swift [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+        PersonDataModel(name: "Lucas"),
+        PersonDataModel(name: "Brian"),
+        PersonDataModel(name: "Mauro"),
+        PersonDataModel(name: "Mugu")
+    ]
+    
+    func getEntities<T: AnyObject>() -> [T] {
+        let result = entities.filter { $0 is T}
+        
+        return result as! [T]
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Persistence/Implementations/InMemoryDataProvider.swift:23]: What happen with this forced casting if result is empty?

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Persistence/Implementations/InMemoryDataProvider.swift [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+//  BadExampleDataProvider.swift
+//  SwiftSeedProject
+//
+//  Created by Mugu on 3/29/17.
+//  Copyright © 2017 Making Sense. All rights reserved.
+//
+
+import Foundation
+
+class InMemoryDataProvider  {
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Persistence/Implementations/InMemoryDataProvider.swift:11]: Extra spacing before curly brace.

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Persistence/Models/PersonDataModel.swift [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+//  SwiftSeedProject
+//
+//  Created by Mugu on 3/29/17.
+//  Copyright © 2017 Making Sense. All rights reserved.
+//
+
+import Foundation
+
+class PersonDataModel : AnyObject {
+    let name: String
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Persistence/Models/PersonDataModel.swift:12]: Please, add a blank line between these two.

File: IOS/SwiftSeedProject/SwiftSeedProject/Interfaces/ViewControllers/Person/PersonTableViewCell.swift [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+//  SwiftSeedProject
+//
+//  Created by Mugu on 3/28/17.
+//  Copyright © 2017 Making Sense. All rights reserved.
+//
+
+import UIKit
+
+class PersonTableViewCell: UITableViewCell {
+    @IBOutlet weak var lblName: UILabel!
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Interfaces/ViewControllers/Person/PersonTableViewCell.swift:12]: Please, add a blank line between these two.

File: IOS/SwiftSeedProject/SwiftSeedProject/Interfaces/ViewControllers/Person/PersonTableViewCell.swift [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+    static let identifier = "PersonCell"
+    
+    override func awakeFromNib() {
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Interfaces/ViewControllers/Person/PersonTableViewCell.swift:15]: Please, remove useless code.

File: IOS/SwiftSeedProject/SwiftSeedProject/Interfaces/ViewControllers/PersonTableViewController.swift [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+//  SwiftSeedProject
+//
+//  Created by Mugu on 3/27/17.
+//  Copyright © 2017 Making Sense. All rights reserved.
+//
+
+import UIKit
+
+class PersonTableViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {
+    @IBOutlet weak var tblView: UITableView!
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Interfaces/ViewControllers/PersonTableViewController.swift:12]: Please, add a blank line between these two.

File: IOS/SwiftSeedProject/SwiftSeedProject/Interfaces/ViewControllers/PersonTableViewController.swift [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+        tblView.dataSource = self
+    }
+
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return (service?.persons.count)!
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: PersonTableViewCell.identifier, for: indexPath) as! PersonTableViewCell
+        cell.lblName.text = service?.persons[indexPath.row].name
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Interfaces/ViewControllers/PersonTableViewController.swift:29]: This line should be handled by the cell class. This view controller can't be responsible of configuring cell's inner views.

**Review by mvazquez-ms — CHANGES_REQUESTED** [review]

File: IOS/SwiftSeedProject/SwiftSeedProject/Domain/Models/Person.swift [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+//  Person.swift
+//  SwiftSeedProject
+//
+//  Created by Mugu on 3/28/17.
+//  Copyright © 2017 Making Sense. All rights reserved.
+//
+
+import Foundation
+
+class Person : NSObject {
```
Comment (briansztamfater) [inline IOS/SwiftSeedProject/SwiftSeedProject/Domain/Models/Person.swift:11]: I think models (and any other Swift class) should not extend NSObject, unless is necessary

**Review by briansztamfater — CHANGES_REQUESTED** [review]

**Pull request closed without merging**